### PR TITLE
New GLSL frontend using pomelo

### PR DESCRIPTION
**Pull request by pjoe**
Work in progress, writing a new GLSL parser using [pomelo.](https://docs.rs/pomelo)

Goal is to reduce output size (especially for wasm), as well as making the parser read better against the spec.

Status:
- [x] port lexer
- [x] syntax parser using pomelo
- [ ] error handling
- [ ] generate parse tree
- [ ] tests

For now done i separate `glsl_new`module.
Also added `PREFER_GLSL_NEW` env var that you can set to 1, to to try this in the convert example.

### CODE REVIEW

File: src/front/glsl_new/error.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+use std::fmt;
+use std::io;
+
+#[derive(Debug)]
+pub enum ErrorKind {
+    InvalidInput,
+    IoError { error: io::Error },
```
Comment (kvark) [inline src/front/glsl_new/error.rs:7]: nit: any reason to have the named field here? seems very straightforward to just have `IoError(io::Error)`

File: src/front/glsl_new/error.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+use std::fmt;
```
Comment (kvark) [inline src/front/glsl_new/error.rs:1]: nit: let's use the include style of other modules, i.e. one statement per top-level use. In this case, `use std::{fmt, io}`

File: src/front/glsl_new/lex.rs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+            self.input = String::from(rest);
+            let meta = token.extra_mut();
+            let end = meta.chars.end;
+            meta.line = self.line;
+            meta.chars.start += self.offset;
+            meta.chars.end += self.offset;
+            self.offset += end;
+            Some(token)
+        } else {
+            match self.lines.next() {
```
Comment (kvark) [inline src/front/glsl_new/lex.rs:280]: nit: can do `let (line, input) = self.lines.next()?;` and proceed without a match

File: src/front/glsl_new/error.rs (hunk 1 above)
Comment (pjoe) [reply src/front/glsl_new/error.rs]: Done

File: src/front/glsl_new/error.rs (hunk 2 above)
Comment (pjoe) [reply src/front/glsl_new/error.rs]: Done

File: src/front/glsl_new/lex.rs (hunk 3 above)
Comment (pjoe) [reply src/front/glsl_new/lex.rs]: Done

File: Cargo.toml [hunk 4]
```diff
@@ -16,6 +16,7 @@ log = "0.4"
 num-traits = "0.2"
 spirv = { package = "spirv_headers", version = "1.4.2", optional = true }
 glsl = { version = "4", optional = true }
+pomelo = "0.1.4"
```
[the remark this answers is not in the document]
Comment (pjoe) [reply Cargo.toml]: Ok, I put everything into a `glsl-new` feature, so all optional now :)

Made the logic in convert example a bit complex, but that will go away once mature enough to replace 'old' glsl front.

**Pull request merged**